### ISSUE TITLE
Replace 'configurationsXyz' and 'modulesXyz' with 'configurations.xyz' and 'modules.xyz'

### ISSUE THREAD

**Issue by gytis-ivaskevicius** [NixOS/nix#6257]
Simply put - the current approach is just not scalable, we got projects defining custom modules/configurations which are being recognized as 'unknown' due to lack of type system.
Some peeps suggest implementing nixpkgs options system or type system via JSON spec but I think the best solution here is to just remove the need for type system entirely.

Expected properties rename:
`nixosConfigurations` to `configurations.<name>`
`nixosModules` to `modules.<name>`
(Similar idea to REST naming convention and existing system-specific properties)

**Comment by gytis-ivaskevicius** [comment]
This allows the community to define its own module groups. Here are a few examples:

- `nix-darwin` users would use `modules.darwin` and `confiugurations.darwin` (currently they use `darwinModules` and `darwinConfigurations` attributes, which are 'unknown' and not checked by `nix flake check`)
- `home-manager` - same idea as `nix-darwin`
- Everyday nixers could make use out of such namespaces as well, for example having groups such as: `modules.{server,desktop,cli,development,etc}`. In fact I do have something similar implemented here https://github.com/gytis-ivaskevicius/nixfiles/blob/master/suites.nix (lines: 3, 19, 32)

**Comment by gytis-ivaskevicius** [comment]
Oh, and just to be clear
> What's the usefulness of having a semantics-free

It is not semantics-free. Current design is `nixosModules = <module>`, this would change definition to `modules.<name> = <module>`

**Comment by bew** [comment]
> It is not semantics-free. Current design is `nixosModules = <module>`, this would change definition to `modules.<name> = <module>`

But there may be other kinds of modules, nixos ones are not the only ones.
(Same for configuration)

**Comment by gytis-ivaskevicius** [comment]
> But there may be other kinds of modules, nixos ones are not the only ones.

Yeah, which is why I am suggesting to make flakes schema more flexible, so it could accommodate nix-darwin/home-manager/etc modules as well.

**Comment by Pacman99** [comment]
Overall a good idea :+1: I think this would cleanup flake's relationship with configuration systems.

Largest question for me is whether `nix flake check` knows how to "check" home-manager/darwin systems? And if it doesn't I presume it would just check `configurations.nixos.*`. 

Also to be clear it would be `modules.<config-system-name>.<name> = <module>` right?

**Comment by gytis-ivaskevicius** [comment]
> Largest question for me is whether nix flake check knows how to "check" home-manager/darwin systems? And if it doesn't I presume it would just check configurations.nixos.*.

The answer is simple: Nope. One of the checks is the derivation path itself, there are a few options:
1. Create some sort of standard like instead of searching for `config.something.toplevel` just have a standard alias something along the lines of `config.drv`
2. Check for derivation only in case of NixOS configs, free pass for everything else :-1: 
3. Hardcode home manager/darwin projects drv paths into `nix` :-1: 

As for the last part - I did not quite get it. Assuming modules are identical I would expect something along the lines of this:
```nix
{self}:{
    modules.nixos = import ./xyz.nix;
    modules.darwin = self.modules.nixos;
    
    # As of right now common approach is: https://github.com/gytis-ivaskevicius/flake-utils-plus/pull/111/files
    nixosModules = import ./xyz.nix;
    darwinModules = self.nixosModules; # unknown properly
}
```

**Comment by marksisson** [comment]
I posted a copy of my flake (#6723) using nixosConfigurations attribute to contain home-manager and darwin configurations - this flake does not have errors when running `nix flake check`, but the extra complexity highlights the fact that the "configurations" abstraction could be better.

**Comment by roberth** [comment]
With a solution like https://github.com/hercules-ci/flake-parts you could remove the hardcoded schema checking in Nix and let the nixpkgs repo define `nixosConfigurations`, let nix-darwin define `darwinConfigurations`, etc.

I'm still surprised how much Nix is hardcoding ecosystem stuff that it has no business knowing about. It's as if we've forgotten that the nix implementation can't be pinned by expressions or lock files. Good thing flakes are experimental so this stuff can still be removed.

**Comment by DavHau** [comment]
I think getting this change in before the stable release of flakes is important.
More and more projects are evolving in the ecosystem that export modules, but there is no suitable output in the current schema.

Examples for these projects:
  - [flake.parts](https://falke.parts)
  - [jupyenv](https://github.com/tweag/jupyenv) (previously jupyterWith)
  - [drv-parts](https://github.com/DavHau/drv-parts)

And of course there are the well known:
   - home-manager
   - nix-darwin
   
In my opinion, the most reasonable schema would to be a 3 levels nested attrset, like:
`modules.<config-system-name>.<name> = <module>`.

@gytis-ivaskevicius Reading the OP it's not quite clear if you are suggesting a 2-levels or 3-levels nested attrset. Maybe you could edit the OP to make that more clear.

**Comment by gytis-ivaskevicius** [comment]
Updated

**Comment by roberth** [comment]
Earlier, we've decided not to check the NixOS modules with flakes anymore, because there was very little practical utility in doing so (as the things that might end up in the attribute tend to be indistinguishable from valid modules), along with it being somewhat flawed architecturally.
 - https://github.com/NixOS/nix/pull/7356#issuecomment-1346202104

To get more use out of this, I think we need to first establish in the module system which things are definitely not modules. There's probably something to be gained there, by making it check against certain things that are known to be highly unlikely to be modules. Ultimately it is (educated) guesswork, so hardcoding that guesswork in Nix doesn't seem like the right thing to do. Together with the architectural argument, I think it'd be better to defer the checking to frameworks (flake.parts, etc), perhaps with a new interface for eval checks that frameworks can hook into. Nix's responsibility is then reduced to accepting and ignoring a `modules` output, documenting its existence, and letting the community evolve the right checks for it.

Similar logic would apply to the `configurations` attribute.

Superficially this might seem like a lazy-but-overengineered solution, but it is the one that allows community expressions to evolve in a forward compatible manner, noting the Nix's logic can not be pinned, unlike community expressions.

**Comment by roberth** [comment]
An example of evolution on the Nixpkgs side is
 - https://github.com/NixOS/nixpkgs/pull/197547
 
Examples of checking behavior that may evolve
 - Adding a `_type` to the result of `evalModules`
 - Rejecting `_type` in module imports and `evalModules` `modules` arguments
 - Consequences of any ideas to make importing modules from flakes a bit more ergonomic (as discussed in the PR or something else). This one's more of an unknown unknowns example, fwiw.

**Comment by Atry** [comment]
Currently the module system in `lib.modules` is called NixOS Module, and the document is included in NixOS manual. However, there are more use case of module system other than NixOS, for example, https://github.com/hercules-ci/flake-parts .

I wonder if we could rename it to Nix Module and move the document to Nixpkgs manual.

**Comment by bew** [comment]
@Atry I think your message is quite orthogonal to this issue, I'd suggest opening a new issue for that instead

**Comment by Atry** [comment]
I agree documenting might belong to another issue, but it is related to this issue, because this issue is about renaming `nixosModules` to `modules.nixos`, potentially allowing for modules other than NixOS Module.

**Comment by Atry** [comment]
I would argue if "config system" really exists. Most NixOS modules can be used as a `perSystem` flake part without any modification, e.g. https://github.com/hercules-ci/flake-parts/issues/74#issuecomment-1513708722

For example, you can create a [flake-parts](http://flake.parts/) project, then configure some Jupyter kernels using `services.jupyter.kernels` by importing  `"${nixpkgs}/nixos/modules/services/development/jupyter"` to your `perSystem` module, then you can launch Jupyter in your project by turning `systemd.services.jupyter.serviceConfig.ExecStart` into a shell script.

After all, NixOS modules are just config file generators. We don't have to run the generated files in `systemd`.

**Comment by roberth** [comment]
@Atry, while this may be possible today, you should expect this to break, unless adequate testing is added in the nixpkgs repo to ensure that this use case keeps working. This essentially what RFC 78 is about. See my [comment](https://github.com/NixOS/rfcs/pull/78#issuecomment-753487600) for an example of how the module system can be leveraged to avoid the unnecessary NixOS-specific complexity in your use case (although it leaves a lot to the imagination).

**Comment by roberth** [comment]
In [this PR](https://github.com/NixOS/nixpkgs/pull/197547), the concept of "config system" or "module system application" is formalized as `class`. By omitting a `_class` declaration in your modules, you can keep them generic.
Using this term, we can define the current convention as

```nix
flake."${class}Configurations"
flake."${class}Modules"
```

This is also mentioned in the module system documentation (but the online manual hasn't updated yet).

The suggested attribute path convention would be:

```nix
flake.configurations.${class}
flake.modules.${class}
```

Furthermore the PR has added `_type = "configuration";` to the values that are meant to go into `configurations.${...}`.

**Comment by roberth** [comment]
Neither naming scheme has a particularly good spot for generic modules.
In the module system, such modules are represented by `_class = null;` / `class = null;`, but `nullModules` or `modules.null` is unpleasant. In the existing convention, `modules` would be suitable, but this blocks off the potential for an unambiguous `modules.${class}` later.
Perhaps a better alternative is to replace `null` by `"generic"`? That way we can have `genericModules` or `modules.generic` without making the naming rules more complicated. I don't like that it slightly deteriorates the class checking data model though; `nullOr str` is most accurate.

**Comment by roberth** [comment]
Let me reply to @blaggacao here

> **More Context on `home-manager`**
> 
> Home Manager users are [fundamentally system-portable](https://github.com/nix-community/home-manager/issues/2161). Advancement here could trigger evolution there.

#### Definitions

If it's portable, then "system" is a parameter, which means it is _not_ a configuration in the sense of the module system.
Or even in general. Looking at the [wiktionary definitions for configuration](https://en.wiktionary.org/wiki/configuration), none of them suggest that there's any parameter that remains, and neither does it suggest that a configuration is not a singular thing.

It's unfortunate that user expectations and the bottom-up definition lead to an overloading of the term "configuration", and they meet in the flake schema.
Maybe it's helpful to have a term for what `configurations.<system>.<name>` would be; perhaps a _configuration family_, which we could define as a set of configurations that derive from roughly the same modules, but need to vary along a parameter, such as the host platform.

#### Simple schema

The practical problem with configuration families in relation to this issue's goal is that they have a different schema compared to a normal attribute set of configurations. If we were to put both in the `configurations` schema, we wouldn't succeed at making the schema more general, but rather we would have produced a set of renames.

It might still be possible to traverse such a heterogeneous attribute set without significant cost based on `x._type or null == "configuration"`, but I wouldn't count on module system applications deferring their assertions to the `config` attribute alone, and even then it wouldn't be possible to traverse it generically without loading the flakes that implement these module system applications.

So in line with the goal of this issue, actually simplifying the schema, I'd rather keep `configurations` simple; just `attrsOf configuration` so to speak.
An application that needs something else can use its own attribute.

#### Modules instead?

As a final note, a different interpretation is that if a Home Manager configuration is truly portable, then "system" is not part of the configuration. Arguably the derivations it builds need to be configured, so there's a hole in that configuration. Such a hole, whether it's a function parameter or an option is representable using a module rather than a configuration.
Perhaps Home Manager should do the same thing that NixOps has to do. A NixOps configuration has many holes because of cloud resource state info, so it can only be represented by a module rather than a configuration. The configuration only exists within nixops' own specific evaluation.

This would look like:

```nix
modules.home-manager.foo = { ... }: {
  # ...
}
```

While I like the simplicity of this, it has some issues
 - It's not clear which modules are "top level" ones that are suitable as the whole input to the module system application.
 - `specialArgs` can't be declared. While I think `specialArgs` should be avoided as much as possible because it is truly non-modular "global state", it should be possible.
 - It's hard to bootstrap a configuration from this: which `lib` to use? How is `pkgs` initialized?
    - applications will handle this differently

These issues seem sufficient to warrant a new flake output attribute containing this info.
The new flake output attribute would immediately solve the first issue. The others need to be handled by the attribute values.
I don't have a good suggestion for a name yet, so I'll just use `partialConfigurations`.

Brainstorming:

Possible attrs:

- `partialConfigurations.<name>.modules`: list of modules as usual; the flake's contribution to the module application's configuration.
- `partialConfigurations.<name>.specialArgs`: extra specialArgs that the module application should add
- `partialConfigurations.<name>.lib`: version of `lib` to use. I don't know if this is necessary or sufficient. Maybe more dependencies should be specified.
- `partialConfigurations.<name>.application`: version of the module system application to use (e.g. the home manager flake)
- `partialConfigurations.<name>.applicationArgs`: any arguments that module system application accepts. This could remove the need for a `.lib` attr

Actually `.application` suggests that we could standardize how module system applications are invoked. Seems a bit out of scope, and rather powerful. Too powerful?
I initially just forgot to add `.<class>` by mistake, but `.application` seems to fix that. The first two attributes seem necessary, but I have yet to make sense of the rest. Maybe just add `.<class>` to solve the problem at hand without getting carried away.
Needs more thought.

**Comment by blaggacao** [comment]
Without going too far, still:

> perhaps a configuration family,

I could see definition of "configuration levels" helpful as a community-provided shared mental model.

This model captures different aspects (or "levels") of exogenous input, such as system, ip, identity, etc.

With the shared knowledge and understanding of euch levels, we can make use of the two fundamental strategies of configuration management: functions (if there's a "hole") and overlays (if its a "variant").

Having fixed points is a really powerful primitive to build such a mental model and avoid impracticalities (such as rendering the change of trunk values like `system` operationally and conceptually "cheap").

**Comment by bb010g** [comment]
Note that since <https://github.com/NixOS/nixpkgs/pull/344688>, Nixpkgs's `lib.modules` suggests this convention.

**Comment by hsjobeki** [comment]
If i may add a few of my thoughts to this

I'd prefer to have `<vendor>.<ressourceType>` 

Where vendor (or context, or whatever you call it) can be imagined: 
- vendor `nixos`
- ressources: `configuration`, `(nixos) modules`, ...

Where a different vendor i.e `flake-parts` might have completely different ressource types mapped to the same "name":
- `modules` (flake-parts modules)
- ...

Where grouping by resource type first means we are mixing different resources types up:

```nix
{
   modules.flake-part = {} # <- those are flake-parts "modules" It is just a coincidence that they use the same name for their ressource.
   modules.nixos = {}       # <- those are nixos modules
}
```

Naming things is very hard and I could rather often name something a `module`. The word `module` only gains its meaning within the vendor context so it should be nested under it, and not above.

**Comment by Atry** [comment]
Currently, the type of flake-parts’ `nixosConfigurations` is `lazyAttrsOf
raw`, and the flake-parts users define a NixOS configuration like this:

```
flake.nixosConfigurations.myHostname = inputs.nixpkgs.lib.nixosSystem { ...
}
```

As a result, `myHostname` is a module system separately evaluated from
flake parts' module system, and `myHostname` must not have any parameter
that remains.

However, if we change the type of flake-parts’s `nixosConfigurations` to
`lazyAttrsOf submodule`, and let it evaluates along with flake-parts'
module system, then `flake.nixosConfigurations.myHostname` can be merged
from more than one definitions, thus it automatically allows for parameters
that remains. Therefore, `nixosConfigurations.myHostname` is already a
partial configuration!

That is to say, the flake specification does not have to define what is a
partial configuration. Instead, a module system to define a flake, such as
flake-parts, should be able to produce partial configuration.

**Comment by Atry** [comment]
Also I found that the current flake's naming is inconsistent with the current module system API's naming.

| flake attribute names | module system types | flake-parts options | remark |
| - | - | - | - |
| nixosConfigurations | lazyAttrsOf raw | flake.nixosConfigurations | separatedly evaluated in flake-parts |
| nixosConfigurations | lazyAttrsOf submodule | N/A | not yet supported in flake-parts |
| nixosModules | lazyAttrsOf deferredModule | flake.nixosModules | |

An easy way to address the inconsistency might be to rename `types.submodule` to `types.evaluation` or `types.evaluatedModule`, and to rename `types.deferredModule` to `types.module`.

**Comment by roberth** [comment]
> If flakes want vendors to define and control their own ecosystems independently, i recommend to go with `vendor.resourceType`
> 
> Where vendor (or context, or whatever you call it) can be imagined:
> 
> * vendor `nixos`
> * ressources: `configuration`, `(nixos) modules`, ...

Let's split this problem into two parts
1. The module system has claimed the term `module`. `moduleSystemModule` would have been more accurate, but also exposes plain that it, all by itself, has a name which is perhaps too generic. I do not believe we should change this though, because
    - The module system has a special role in the ecosystem as the dominant configuration DSL,
    - It has had this name for decades and renaming it to something else would be disruptive.
2. The module system's `class` feature only supports "flat" strings, as does the `modules.<class>.<moduleName>` convention/"schema".
    - This does not preclude the addition of a vendor prefix, and we could impose a convention through the module system documentation for `class`. Example: `modules."hercules-ci/arion/project".dns-infra = ./arion/dns-infra.nix;` I have no idea whether `/` is the right choice. `_` instead would not need quotes. Can debate in docs update.
    - A structured named (e.g. list of strings) would add no real extra functionality, but would be slower in the interpreter

**Comment by roberth** [comment]
> the inconsistency might be to rename `types.submodule` to `types.evaluation` or `types.evaluatedModule`,

We already have
```nix
let fooConfiguration = evalModules { class = "foo"; };
in assert fooConfiguration._type == "configuration";
```

So an evaluated _set of_ modules is called a _configuration_. You're right that `submodule` is named wrong. I would propose to call it a `subconfiguration` (and reconsider various details while introducing this new name alongside `submodule`).

Rename `types.deferredModule` to `types.module`. I'd be on board with that when `subconfiguration` becomes a thing. In that case there's no need to explicitly differentiate `module` from `submodule` anymore.

**Comment by Atry** [comment]
Why not just `types.configuration`? The fact that a configuration is an attribute of another object should not affect the type of the configuration.

**Comment by roberth** [comment]
It does not provide the full `_type = "configuration";` object. A `submodule`/`subconfiguration` does not return the attribute `options` for example, so it is different.
My addition of a `types.configuration` that returns `_type = "configuration";` was rejected in Nixpkgs because this type is never used within module system applications. `flake-parts` is an exception because it models flakes, which is a rare _domain_ that happens to include things like `nixosConfiguration`. No other domain that has been modeled with the module system contains it, as far as I know.

Side note: https://github.com/NixOS/nixpkgs/pull/391544 may provide _access_ to `options`, but does not make it part of the _value_ (type value or option value &ndash; `removeAttrs config ["_module"]`).

**Comment by Atry** [comment]
How about `types.configurationConfig` or simply `types.config`? The term `config` comes from `(evalModules { ... }).config`.
